### PR TITLE
Implement chart/graph drawing

### PR DESCRIPTION
**Pull request by AaronVanGeffen**
This implements the `CompanyList::drawGraph` function. To maintain sanity, I've spun off some of the logic into separate functions.

During development, I built in a mechanism to quickly switch to the vanilla `drawGraph` function when the shift is held down. The current implementation should be a pixel-perfect match for the vanilla code.

In a future refactor, `graphGetMaxValue` and `drawGraphLineSegments` should be templated to avoid all the manual pointer mess. A special case is likely needed for values dealing with `currency48_t`, as it's the only type that currently cannot be implicitly converted to an `int64_t`.

### CODE REVIEW

File: src/OpenLoco/src/Windows/CompanyList.cpp
Comment (AaronVanGeffen) [inline src/OpenLoco/src/Windows/CompanyList.cpp:1865]: Should we just implement this, keeping support for negative numbers? Alternatively, drop support for this flag entirely? Who knows when it might come in handy, though — perhaps for future plugins?

Comment (duncanspumpkin) [reply src/OpenLoco/src/Windows/CompanyList.cpp:1865]: i'd be more inclined to just delete it for now

File: src/OpenLoco/src/Ui/Chart.cpp
Comment (duncanspumpkin) [inline src/OpenLoco/src/Ui/Chart.cpp:79]: can we name this?

File: src/OpenLoco/src/Ui/Chart.cpp
Comment (duncanspumpkin) [inline src/OpenLoco/src/Ui/Chart.cpp:85]: do we know enough to name word_113DD80

File: src/OpenLoco/src/Ui/Chart.cpp
Comment (duncanspumpkin) [inline src/OpenLoco/src/Ui/Chart.cpp:262]: can remove this

File: src/OpenLoco/src/Ui/Chart.h
Comment (duncanspumpkin) [inline src/OpenLoco/src/Ui/Chart.h:26]: Do we still need to pack?

File: src/OpenLoco/src/Windows/CompanyList.cpp
Comment (duncanspumpkin) [inline src/OpenLoco/src/Windows/CompanyList.cpp:34]: does this need to be a loco_global still?

File: src/OpenLoco/src/Windows/CompanyList.cpp
Comment (duncanspumpkin) [inline src/OpenLoco/src/Windows/CompanyList.cpp:1737]: highlighted

File: src/OpenLoco/src/Ui/Chart.cpp
Comment (AaronVanGeffen) [reply src/OpenLoco/src/Ui/Chart.cpp:79]: Something like xTickOffset?

File: src/OpenLoco/src/Ui/Chart.cpp
Comment (AaronVanGeffen) [reply src/OpenLoco/src/Ui/Chart.cpp:85]: We already have `yAxisLabelIncrement`, so this would be `xAxisLabelIncrement`.

File: src/OpenLoco/src/Windows/CompanyList.cpp
Comment (AaronVanGeffen) [reply src/OpenLoco/src/Windows/CompanyList.cpp:34]: Good catch! No longer necessary indeed.

File: src/OpenLoco/src/Ui/Chart.h
Comment (AaronVanGeffen) [reply src/OpenLoco/src/Ui/Chart.h:26]: We won't if we let go of the loco_global, as you suggested. Great! Also means the enum flags no longer need to be uint32_t.

File: src/OpenLoco/src/Ui/Chart.cpp
Comment (AaronVanGeffen) [inline src/OpenLoco/src/Ui/Chart.cpp:279]: What would be a good name for this?

I think it's a bit strange how it alternates between drawing points and lines, tbh. Initially, I thought it would make the points more visible for lines with overlap, but that doesn't appear to be the case.

Comment (AaronVanGeffen) [reply src/OpenLoco/src/Ui/Chart.cpp:279]: Actually, it seems `byte_113DD99` is only ever set to 1, so the points (flag 0) are never even drawn...

File: src/OpenLoco/src/Economy/Economy.cpp
Comment (duncanspumpkin) [inline src/OpenLoco/src/Economy/Economy.cpp:81]: can't we return a proper type like `const std::array<currency32_t, 60>&`

File: src/OpenLoco/src/Windows/CompanyList.cpp
Comment (duncanspumpkin) [inline src/OpenLoco/src/Windows/CompanyList.cpp:1129]: if we used a proper type for getDeliveryCargoPaymentsTable this could then use the `size()` of that proper type

File: src/OpenLoco/src/Ui/Chart.h
Comment (duncanspumpkin) [inline src/OpenLoco/src/Ui/Chart.h:44]: maybe we should consider a constant for the 32 max number of lines

File: src/OpenLoco/src/Economy/Economy.cpp
Comment (AaronVanGeffen) [reply src/OpenLoco/src/Economy/Economy.cpp]: I've gone with a `std::span` instead. Seemed appropriate.

File: src/OpenLoco/src/Ui/Chart.h
Comment (LeftofZen) [inline src/OpenLoco/src/Ui/Chart.h:64]: did you mean "powers of 2"?

**Review by LeftofZen — CHANGES_REQUESTED** [review]
I checked the pixels and its so close, I notice in your that some parts of the line are thicker (these are the 'lighter' pixels here, seen under a subtract filter in photoshop)

image

Edit: actually I just ran the same graph again in yours and master both, and

File: src/OpenLoco/src/Ui/Chart.h
Comment (AaronVanGeffen) [reply src/OpenLoco/src/Ui/Chart.h:64]: The comment is a bit dated -- it stems from back when the variable wasn't named. It's shifts, so powers of two by definition. Suppose we can just remove the comment at this point.

**Comment by AaronVanGeffen** [comment]
> I checked the pixels and its so close, I notice in your version that some parts of the line are thicker (these are the 'lighter' pixels here, seen under a subtract filter in photoshop)
>
> I ran it a second time just to be sure, and the exact same pixels are different, so its deterministic. Judging by the error, I'd say its an off-by-one somewhere?

Hmm, that does suggest another deviation in our `drawLine` algorithm, indeed — see also #2657. To be honest, I feel like it's so minor that I don't want to hold back this PR for it. We should definitely file an issue for it though.

**Comment by LeftofZen** [comment]
> To be honest, I feel like it's so minor that I don't want to hold back this PR for it. We should definitely file an issue for it though.

This sounds perfectly reasonable to me, happy for that to happen.

**Review by LeftofZen — APPROVED** [review]

**Comment by AaronVanGeffen** [comment]
I had forgotten I had enabled auto-merge for this PR. Fully intended to wait for @duncanspumpkin to approve this before merging. Sorry for the oversight.

**Comment by duncanspumpkin** [comment]
> I had forgotten I had enabled auto-merge for this PR. Fully intended to wait for @duncanspumpkin to approve this before merging. Sorry for the oversight.

Oh it's fine i was okay with it.

**Pull request merged**
